src/domain/core/src/entities/metadata_chain.rs

```diff
@@ -185,15 +192,21 @@ pub trait MetadataChainExt: MetadataChain {
         self.accept_by_hash(visitors, &head_hash).await
     }
 
+    /// A lower-level method of accepting Visitors ([MetadataChainVisitor])
+    /// to specify their initial decisions.
+    ///
+    /// See also [Self::accept()]
     async fn accept_by_hash_with_decisions<E>(
         &self,
-        decisions: &mut [Decision],
+        decisions: &mut [MetadataVisitorDecision],
         visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
```

Comment:
> Since these two are so interconnected logically, shouldn't it be wrapped in a struct, something like `VisitingState` ? (c) https://github.com/kamu-data/kamu-cli/pull/523#discussion_r1517496388
>
> If you decide to add a counter of stopped visitors, it would logically belong to this struct very naturally. (c) https://github.com/kamu-data/kamu-cli/pull/523#discussion_r1517497046

Reply:
I had a similar solution earlier (maybe with a less successful name than the one you suggested), but it was rejected.

I think we can leave it as it is, since these methods contain all the logic in themselves, and it's simple.

Reply:
The "VisitorFacade" you had before did not perform any additional logic or held any state, thus to me it was only obscuring what is going on in accept loop and I suggested removing it.

Space for creativity is big here, but let's be driven by real use cases and keep the solution no more complex than needed to satisfy them.

Reply:
> The "VisitorFacade" you had before did not perform any additional logic or held any state, thus to me it was only obscuring what is going on in accept loop and I suggested removing it.

I meant that `VisitorFacade` would accept the "initial" decisions and `Visitors` as well as the proposed variant.

but yes, of course, it had no additional state

> Space for creativity is big here, but let's be driven by real use cases and keep the solution no more complex than needed to satisfy them.

:+1: